---
title: "Class 15: Genome Informatics II"
author: "Serina Huang"
date: "`r format(Sys.time(), '%D')`"
output: github_document
---

## Transcriptomics and RNA-seq Analysis
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
# source("http://bioconductor/org/biocLite.R")
# biocLite()
# biocLite("DESeq2")
# library(BiocInstaller)

# stringsAsFactors reads a logical. If true, charater vectors will be converted to factors
counts <- read.csv("https://bioboot.github.io/bimm143_W18/class-material/airway_scaledcounts.csv", stringsAsFactors = FALSE)
metadata <- read.csv("https://bioboot.github.io/bimm143_W18/class-material/airway_metadata.csv", stringsAsFactors = FALSE)

head(counts)
head(metadata)
```

### Rudimentary Differential Gene Expression Analysis
Let's examine the metadata file to find __control__ and __treated__ columns. Then, compute the mean counts per gene for the each treatment.
```{r}
# Find control samples in metadata
control_ind <- metadata$dex == "control"
# Narrow down metadata to only control samples
control <- metadata[control_ind,]

# Compute mean of each gene for control samples
control.mean <- rowSums(counts[,control$id]) / nrow(control)
# Assign gene ID to each mean gene count
names(control.mean) <- counts$ensgene
```

Do the same thing for the treated columns:
```{r}
# Find treated samples in metadata
treated_ind <- metadata$dex == "treated"
treated <- metadata[treated_ind,]

# Compute mean counts per gene across all the treated samples
treated.mean <- rowSums(counts[,treated$id]) / nrow(treated)
head(treated.mean)
# Assign gene ID to each mean gene count
names(treated.mean) <- counts$ensgene
head(treated.mean)
```

```{r}
# Combine control and treated mean count data
meancounts <- data.frame(control.mean, treated.mean)
head(meancounts)
colSums(meancounts)
```

There seems to be differential gene expression between the control and treated groups. However, this is a rudimentary analysis because we did not account for sequencing depth. I.e. the number of reads generated from a sample in a given experiment varies. The deeper the sample is sequenced, the more transcripts are identified, and the higher the expression level.

Let's plot this result before moving on to the real differental gene analysis.
```{r}
plot(meancounts$control.mean, meancounts$treated.mean, xlab = "Control", ylab = "Treated")
```

Each point is a gene. The points off the diagonal are the genes with differential expression. There is a lot of data with low counts off the diagonal but they are covering up each other, so let's plot the same data on a log scale.
```{r}
plot(meancounts$control.mean, meancounts$treated.mean, log = "xy", xlab = "log Control", ylab = "log Treated")
```

We usually quantify this difference using log~2~ fold change since __doubling__ is either to understand.
```{r}
# Compute log2 fold change and add the results to our dataframe
meancounts$log2fc <- log2(meancounts[,"treated.mean"] / meancounts[,"control.mean"])
head(meancounts)
```

We get NaN if we divide by 0 (i.e. the control has 0 mean gene expression for a particular gene) and try to take the log. We get -Inf if we take the log of 0 (i.e. the treated has 0 mean gene expression for a particular gene).

Let's filter out the genes with zero expression.
```{r}
# For each gene, determine if the mean count is 0
# Do so for both control and treated samples
x <- meancounts[,1:2] == 0
head(x)

# Identify the index of each gene for which BOTH control and treated samples are 0 
zero.vals <- which(x, arr.ind = TRUE)
head(zero.vals)

# Store the indices of the genes that we want to remove
to.rm <- unique(zero.vals[,1])
# The negative sign means: remove this index from the array
mycounts <- meancounts[-to.rm,]
head(mycounts)

# Now we can see which genes are upregulated or downregulated
up.ind <- mycounts$log2fc > 2
down.ind <- mycounts$log2fc < (-2)

# How many genes are upregulated or downregulated?
sum(up.ind)
sum(down.ind)
# How many genes in total have differential expression between control and treated groups?
sum(up.ind) + sum(down.ind)
```

Annotate our genes so we know what genes were have and might be doing.
```{r}
anno <- read.csv("https://bioboot.github.io/bimm143_W18/class-material/annotables_grch38.csv")
head(anno)
```

Let's use the `merge()` function to add the annotation data to our our `mycounts` data frame.
```{r}
# by.x = column name in x to merge on
# by.y = column name in y to merge on
# All data frames have a row.names attribute
mycounts.anno <- merge(mycounts, anno, by.x = "row.names", by.y = "ensgene")
head(mycounts.anno)
```

However, if you don't have an annotation file, you can use the following Bioconductor package:
```{r}
library(AnnotationDbi)
library(org.Hs.eg.db)
# See all available keytypes
columns(org.Hs.eg.db)
```

The `multiVals` argument tells the function what to do if there are multiple possible values for a single input value. Here, we ask for the first one that occurs in the database.
```{r}
# Another way to do the same thing
mycounts$symbol <- mapIds(org.Hs.eg.db,
                     keys = row.names(mycounts),
                     column = "SYMBOL",
                     keytype = "ENSEMBL",
                     multiVals = "first")
head(mycounts)
```

```{r}
mycounts$entrez <- mapIds(org.Hs.eg.db,
                      keys = row.names(mycounts),
                      column = "ENTREZID",
                      keytype = "ENSEMBL",
                      multiVals = "first")

mycounts$uniprot <- mapIds(org.Hs.eg.db,
                      keys = row.names(mycounts),
                      column = "UNIPROT",
                      keytype = "ENSEMBL",
                      multiVals = "first")
head(mycounts)
```

### DESeq2 Analysis
```{r}
library(DESeq2) 
dds <- DESeqDataSetFromMatrix(countData = counts,
                              colData = metadata,
                              design =~ dex,
                              tidy = TRUE)

# The DESeq function normalizes library size by estimating "size factors", estimates dispersion for the negative binomial model, and fits models and gets statistics for each gene
dds <- DESeq(dds)
res <- results(dds)
res
summary(res)
```

We see that the default p-value and LFC thresholds are pretty lenient, so let's change the p-value (alpha) to 0.05
```{r}
res05 <- results(dds, alpha = 0.05)
summary(res05)

res01 <- results(dds, alpha = 0.01)
summary(res01)
```

Add annotations to our results object `res01`.
```{r}
head(res01)
res01$symbol <- mapIds(org.Hs.eg.db,
       keys = row.names(res01),
       column = "SYMBOL",
       keytype = "ENSEMBL",
       multiVals = "first")

# Order results by p-value or magnitude (fold change)
ord <- order(res01$padj)
head(res01[ord,])
# Write out the ordered significant results with annotations
write.csv(res01, "signif01_results.csv")
```

## Data Visualization
Let's focus on the gene CRISPLD2 as in the paper.
```{r}
# First, find the gene ID for CRISPLD2
# grep looks for CRISPLD2 inside the res01$symbol vector
i <- grep("CRISPLD2", res01$symbol)
i
rownames(res[i,])
```

Note: Fold change is not too big for CRISPLD2, but very small p-value.

### Box Plots
Let's make a box plot, with dex treatment being our `intgroup`, or interesting group variable.
```{r}
plotCounts(dds, gene = "ENSG00000103196", intgroup = "dex")
```

This looks mediocre. Let's return the results of `plotCounts` to make our own box plot.
```{r}
d <- plotCounts(dds, gene = "ENSG00000103196", intgroup = "dex", returnData = TRUE)
d
```

We see that dex-treated samples have a higher expression of the CRISPLD2 gene.
```{r}
# Split count based on dex option
boxplot(count ~ dex, data = d)
```

We can make this box plot prettier using `ggplot2`.
```{r}
library(ggplot2)
ggplot(d, aes(dex, count)) + geom_boxplot(aes(fill=dex)) + scale_y_log10() + ggtitle("CRISPLD2")
```

### Volcano Plots
In more recent papers, volcano plots have become popular since they show fold change and the p-value. Let's add a `sig` column to our `res` results that evaluates __TRUE__ if padj < 0.05, __FALSE__ if not, and __NA__ if padj = NA.
```{r}
res$sig <- res$padj < 0.05
# How many of each?
table(res$sig)
# How many have padj = NA?
sum(is.na(res$sig))
```

On our volcano plot, color the gene if it has differential expression.
```{r}
ggplot(as.data.frame(res), aes(log2FoldChange, -1*log10(pvalue), col=sig)) + geom_point() + ggtitle("Volcano plot")
```

The following code also makes a Volcano plot without the `ggplot2` library.
```{r}
plot(res01$log2FoldChange, -log(res01$padj))
```

Make a color vector to highlight the points we want (i.e. genes with high fold change and small p-values).
```{r}
mycols <- rep("black", nrow(res01))

mycols_ind <- (abs(res01$log2FoldChange) > 2) & (res01$padj < 0.01)

mycols[mycols_ind] <- "red"

plot(res01$log2FoldChange, -log(res01$padj), col = mycols)

```

------

### Reference
Himes _et al_. RNA-Seq transcriptome profiling identifies CRISPLD2 as a glucocorticoid responsive gene that modulates cytokine function in airway smooth muscle cells. _PLoS One_ 2014;9(6):e99625.