---
title: "Class 14: Genome Informatics I"
author: "Serina Huang"
date: "`r format(Sys.time(), '%D')`"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Section 1. Identify SNPs associated with Childhood Asthma Using OMIM

Q1. What are the 4 candidate SNPs that demonstrate significant evidnece for association to childhood asthma?

A1. After searching for `Verlaan childhood asthma` on OMIM, we found that Verlaan _et al._ reported 4 SNPs associated to childhood asthma risk: rs12936231, rs8067378, rs9303277, and rs7216389.

Q2. What 3 genes do these variants overlap or effect?

A2. ZPBP2, GSDMB, and ORMDL3.

Let's explore the location of these 4 SNPs and 3 genes. On the Ensembl website, we can find the coordinates of the SNPs and whether each SNP overlaps the genes or is in between genes (i.e. intergenic).

Q3. What is the location of rs8067378? What are the different alleles the SNP can have?

A3. After searching rs8067378, we found that it is located on Chromosome 17:39,895,095, on the forward strand. The variants are A|G, with G having a frequency of 43%.

Q4. What are the downstream genes for rs8067378?

A4. On Ensemble -> Phenotype data, we found that the downstream genes of this variant include GSDMB, ORMDL3, ZPBP2, and IKZF3.

Now, let's check out the genotypes of this variant in the Mexican Ancestry in Los Angeles, California (MXL) population. Go to Ensembl -> Sample genotypes.

```{r}
# Examine asthma SNPs in the MXL (Mexican Ancestry in Los Angeles, California) 1000 Genomes sequencing data
mxl <- read.csv("373531-SampleGenotypes-Homo_sapiens_Variation_Sample_rs8067378.csv")
```

Q5. What proportion of the MXL population are homozygous for the asthma associated SNP (G|G)?

A5. After checking out `mxl`, we see that the second column contains genotype info. We can tabulate these genotypes.
```{r}
table( mxl[,2] )
```

This means that AA is the most prevalent genotype, with 22 people. What percentage is this?
```{r}
genotypes <- round( table(mxl[,2]) / nrow(mxl) * 100 )
genotypes
```

There are `r genotypes[4]`% of AA genotype in this population.

Q6. Back on Enembl, search for the particular sample HG00109. This is a male from the British in England and Scotland (GBR) population. What is the genotype for this sample?

A6. G|G.

### Section 2. Initial RNA-seq Analysis

In this section, we want to understand whether the G|G SNP will affect gene expression. We downloaded the modified (trimmed) RNA-seq files for the HG00109 from the class website: [File 1](https://bioboot.github.io/bimm143_F18/class-material/HG00109_1.fastq) and [File 2](https://bioboot.github.io/bimm143_F18/class-material/HG00109_2.fastq).

Let's examine these files in Unix using `head HG00109_1.fastq`, `tail HG00109_1.fastq`, or `less HG00109_1.fastq`.

We noticed that there are 4 lines per sequence in a FASTQ file. For example:
```
@HW-ST546:136:D0HWFACXX:5:1101:10675:54538
TTTCTGTCACAGCTTCCCCATACCCCCGCAGGAGTCAGGGCCAAGTTGGACCTGTGCTTACAATGACTCACAGGC
+
DDDDCDEDCDDDDBBDDDCC@=DHIIGIIFJJJIGJIJJJJJJJJJJJJHJJIGIGJJIJIIHHHFHFFFFFCCC
@HW-ST546:136:D0HWFACXX:5:1101:10891:51711
CAGCTGTCGGCTTCAAACAGACGGTGCACAAGCTTCTCATCGTGGTGGAGCTGTTGCTGGGGGAGATCCCTGACC
+
CCCFFFFFHHHGHJJJJIJIJJJJGGBHIIIJJJJJIIDGHGGGGCGG@FHIIIEIGIHHHHF8>3;CDDAACCA
```

Line 1: begins with @ followed by a sequence identifier and (optional) description.
Line 2: raw nucleotide sequence.
Line 3: begins with + and (optional) same sequence identifier again.
Line 4: quality value for each nucleotide.

To avoid installing a bunch of packages on our laptop, we will do the following analyses on a Galaxy server. To access our Galaxy Server, paste IP into url and register as a user.

Upload HG00109_1.fastq and HG00109_2.fastq to Galaxy.

Q7. How many sequences are in the first file? What is the file size and format of the data?

A7. 3,863 sequences. 741.9 KB and fastqsanger format.

We can't do a t-test or simple statistics to analyze sequencing data. For example, we might get 20,000 reads. 5% confidence for 20,000 reads is really not enough for us to be confident.

Q8. Does the first sequence have good quality?

A8. We can look at the quality score for each nucleotide in R.
```{r}
# Remove comment if packages have not been installed
# install.packages("seqinr")
# install.packages("gtools")
library(seqinr)
library(gtools)

# Convert ASCII characters to Phred quality scores
phred <- asc( s2c("DDDDCDEDCDDDDBBDDDCC@")) - 33
phred

# Phred quality scores are linked to error probabilities on a logarithmic scale
# Q = -10 log P
# Where Q is the Phred quality score and P is the base-calling error probability
# For example, a Phred quality score of 10 means that 1 in 10 bases was called incorrectly, i.e. base call accurary of 90%
prob <- 10**(-phred/10)
prob
prob_percent <- round( prob * 100, 2 )
prob_percent
```

What does this mean? The first base that was called had a quality score of `r phred[1]`, which converts to a `r prob_percent[1]`% that the base was incorrectly called. That's pretty good! The last base, on the other hand, had a quality score of `r phred[ length(phred) ]`, which coverts to a `r prob_percent[ length(phred) ]`% that the base was incorrectly called.

Usually, the ends of reads have lower quality scores, so we typically trim the ends.

Now, let's do this on Galaxy so we can look at the entire data.
```
NGS: QC and manipulation -> FastQC -> Execute
```

This is equivalent to doing `fastqc filename.fastq` on the command line. View data of Webpage and check quality. In the first plot of basic statistics, the x-axis is titled __Position in read (bp)__; it goes from 1 to 75. This makes sense considering that each read has 75 bases in it when we inspected the head of `HG00109_1.fastq`. Note: Anything in the red area should not be used in analysis.

Q9. What is the GC content and sequence length of the second fastq file?

A9. Under `Per sequence GC content`, we see that over 450 sequences had approximately 54% GC. The red curve above the blue curve also indicates that we had more GC content than theoretical expectation. Under `Sequence Length Distribution`, we see that almost all 3,683 sequences had a sequence length of 75 bp. Note: `Per base N content should be a flat line.`

Q10. How about per base sequence quality? Does any base have a median quality score below 20?

A10. Blue line is the mean quality score. For this exercise, let's assume that a median quality score < 20 is unusable. All bases have a median quality score > 20, so there's no need to trim the dataset.

### Section 3. Map RNA-seq Reads to Genome

Now, we want to map the processed reads to the genome. The challenge in mapping RNA-seq reads is that the reads often cross splice junctions (i.e. alternative splicing of exons). Thus, typical NGS mappers like __Bowtie__ or __BWA__ would not be suitable. Instead, __TopHat__ is designed to map RNA-seq by handling gaps.
```
NGS -> TopHat -> Execute
```

The program outputs an `accepted_hits` file that is in __BAM__ format, this is binary so we need to convert it to __SAM__ to inspect the results.
```
NGS: SAMtools -> BAM-to-SAM
```

Now, we can inspect the alignment on the UCSC Genome Browser.

Q11. Where are the most accepted hits located?

A11. The reads pile up at the GSDMB and PSMD3 genes when we narrow the window to the region __chr17:38007296-38170000__. However, it's cumbersome and infeasible to check every single gene. Cufflinks can helps us with that if we tell the program where the boundaries of each gene are, and then the program can count up how many reads are mapped to each gene.

```
NGS: RNA Analysis -> Cufflinks
```

Use `gene.chr17.gtf` as our reference annotation file, which contains information on the locations of exons, stop codons, etc.

Q13. Cufflinks again produces multiple output files that you can inspect on the history sidebar. From the “gene expression” output, what is the FPKM for the
ORMDL3 gene? What are the other genes with above zero FPKM values?

A13. 136,853.

### Section 4. Population Scale Analysis

Switching gears, we want to assess genetic differences on a population scale. In this section, we will determine if there is any association of the 4 asthma SNPs to ORMDL3 expression.

In other words, does having a different genotype at rs8067378 contribute to a higher expression of the ORMDL3 gene?
```{r}
# Load the dataset
ormdl3 <- read.table("https://bioboot.github.io/bimm143_S18/class-material/rs8067378_ENSG00000172057.6.txt")

head(ormdl3)
# We see that the first column is the sample name
# The second column is the genotype
# The third column is the expression value (normalized across samples)
```

Q14. What is the sample size for each genotype? What is the corresponding median expression levels for each genotype?
```{r}
summary(ormdl3)

inds_aa <- ormdl3$geno == "A/A"
summary(ormdl3$exp[inds_aa])

inds_ag <- ormdl3$geno == "A/G"
summary(ormdl3$exp[inds_ag])

inds_gg <- ormdl3$geno == "G/G"
summary(ormdl3$exp[inds_gg])
```

A14. There are 108 samples for the A/A genotype, 233 for A/G, and 121 for G/G. We see that the median expression value for each genotype _is_ different, i.e. the G/G phenotype has lower expression.

Let's put all of this summary in a plot because the median might not be indicative of real difference among the genotypes. We want to display the range, median, and quartiles.
```{r}
# boxplot(count ~ spray, data = InsectSprays, col = "lightgray")
boxplot(exp ~ geno, data = ormdl3, col = "lightgray", xlab = "Genotype", ylab = "Gene Expression", main = "ORMDL3 Exp at rs8067378")
```

Q15. What could you infer from the relative expression value between A/A and G/G? Does the SNP affect the expression of ORMDL3?

A15. A genotype of A/A has a higher expression of ORMDL3 than G/G. The SNP most likely affects the expression of ORMDL3.

------

### Recap

Section 1. We found 4 SNPs associated with childhood asthma from OMIM, checked their target genes on Ensembl, calculated the proportion of people in a population with a certain genotype, and found the genotype of a specific DNA sample.

Section 2. We learned about FASTQ, which is the format of raw RNA-seq data.

Section 3. We uploaded the FASTQ files on Galaxy for downstream analyses. We assessed the quality of the data using Phred quality score, GC content, per base sequence quality, and other more complex statistical methods. We aligned the reads to the reference genome for human using TopHat (vs. other methods). To visualize the alignment on the UCSC Genome Browser, we converted BAM to SAM format. To get a comprehensive view of what genes the reads align to, we used Cufflinks to map the piled up reads to all genes.

Section 4. We analyzed differential expression of the ORMDL3 gene at a given variant in a population using R.

------

### Reference
Verlaan _et al_. Allelic-specific chromatin remodeling in the ZPBP2/GSDMB/ORMDL3 locus associated with the risk of asthma and autoimmune disease. _Am. J. Hum. Genet._ 2009;85:377-393.
